src/ray/gcs/gcs_server/gcs_node_manager.cc

```diff
@@ -19,100 +19,55 @@
 
 namespace ray {
 namespace gcs {
-GcsNodeManager::GcsNodeManager(boost::asio::io_service &io_service,
-                               gcs::NodeInfoAccessor &node_info_accessor,
-                               gcs::ErrorInfoAccessor &error_info_accessor)
+
+GcsNodeManager::NodeFailureDetector::NodeFailureDetector(
+    boost::asio::io_service &io_service, gcs::NodeInfoAccessor &node_info_accessor)
     : node_info_accessor_(node_info_accessor),
-      error_info_accessor_(error_info_accessor),
       num_heartbeats_timeout_(RayConfig::instance().num_heartbeats_timeout()),
-      heartbeat_timer_(io_service) {
-  Start();
+      detect_timer_(io_service) {
+  Tick();
 }
 
-void GcsNodeManager::HandleHeartbeat(const ClientID &node_id,
-                                     const rpc::HeartbeatTableData &heartbeat_data) {
-  heartbeats_[node_id] = num_heartbeats_timeout_;
-  heartbeat_buffer_[node_id] = heartbeat_data;
+void GcsNodeManager::NodeFailureDetector::AddNode(const ray::ClientID &node_id) {
+  heartbeats_.emplace(node_id, num_heartbeats_timeout_);
 }
 
-void GcsNodeManager::Start() {
-  RAY_LOG(INFO) << "Starting gcs node manager.";
-  const auto lookup_callback =
-      [this](Status status, const std::vector<rpc::GcsNodeInfo> &node_info_list) {
-        for (const auto &node_info : node_info_list) {
-          if (node_info.state() != rpc::GcsNodeInfo::DEAD) {
-            // If there're any existing alive clients in client table, add them to
-            // our `heartbeats_` cache. Thus, if they died before monitor starts,
-            // we can also detect their death.
-            // Use `emplace` instead of `operator []` because we just want to add this
-            // client to `heartbeats_` only if it has not yet received heartbeat event.
-            // Besides, it is not necessary to add an empty `HeartbeatTableData`
-            // to `heartbeat_buffer_` as it doesn't make sense to broadcast an empty
-            // message to the cluster and it's ok to add it when actually receive
-            // its heartbeat event.
-            heartbeats_.emplace(ClientID::FromBinary(node_info.node_id()),
-                                num_heartbeats_timeout_);
-          }
-        }
-        Tick();
-      };
-  RAY_CHECK_OK(node_info_accessor_.AsyncGetAll(lookup_callback));
+void GcsNodeManager::NodeFailureDetector::HandleHeartbeat(
+    const ClientID &node_id, const rpc::HeartbeatTableData &heartbeat_data) {
+  auto iter = heartbeats_.find(node_id);
+  if (iter == heartbeats_.end()) {
+    // Ignore this heartbeat as the node is not registered.
```

Comment:
do we know when would this happen?

Reply:
> do we know when would this happen?

If the detector detects that the node is dead due to network reasons, then the node will be removed from `heartbeats_, and then suddenly the `Detector` receives the heartbeat of the removed node again, then it should be ignored.

Reply:
In this case, maybe we should reply the raylet with an error. So the raylet can crash itself as soon as possible. 
Currently, the raylet would crash itself only after receiving the pub sub event.
This is doesn't need to be addressed in this PR. But I think we can add a TODO here.